velox/exec/tests/AggregationFuzzer.cpp

```diff
@@ -0,0 +1,640 @@
+  if (!planPathOpt.has_value()) {
+    planPath = "Failed to create file for saving SQL.";
+  } else {
+    planPath = planPathOpt.value();
+    try {
+      saveStringToFile(
+          plan->toString(true /*detailed*/, true /*recursive*/),
+          planPath.c_str());
+    } catch (std::exception& e) {
+      planPath = e.what();
+    }
+  }
+
+  LOG(INFO) << "Persisted input: " << inputPath << " and plan: " << planPath;
+}
+
+CallableSignature AggregationFuzzer::pickSignature() {
+  size_t idx = boost::random::uniform_int_distribution<uint32_t>(
+      0, signatures_.size() + signatureTemplates_.size() - 1)(rng_);
+  CallableSignature signature;
+  if (idx < signatures_.size()) {
+    signature = signatures_[idx];
+  } else {
+    const auto& signatureTemplate =
+        signatureTemplates_[idx - signatures_.size()];
+    signature.name = signatureTemplate.name;
+    velox::test::ArgumentTypeFuzzer typeFuzzer(
+        *signatureTemplate.signature, rng_);
+    VELOX_CHECK(typeFuzzer.fuzzArgumentTypes(FLAGS_max_num_varargs));
+    signature.args = typeFuzzer.argumentTypes();
+  }
+
+  return signature;
+}
+
+void AggregationFuzzer::go() {
+  VELOX_CHECK(
+      FLAGS_steps > 0 || FLAGS_duration_sec > 0,
+      "Either --steps or --duration_sec needs to be greater than zero.")
+
+  auto startTime = std::chrono::system_clock::now();
+  size_t i = 0;
+
+  while (!isDone(i, startTime)) {
+    LOG(INFO) << "==============================> Started iteration " << i
+              << " (seed: " << currentSeed_ << ")";
+
+    // Pick a random signature.
+    CallableSignature signature = pickSignature();
+    statistics_.functionNames.insert(signature.name);
+
+    std::vector<TypePtr> argTypes = signature.args;
+    std::vector<std::string> argNames = makeNames(argTypes.size());
+    auto call = makeFunctionCall(signature.name, argNames);
+
+    // 20% of times use mask.
+    std::vector<std::string> masks;
+    if (vectorFuzzer_.coinToss(0.2)) {
+      ++statistics_.numMask;
+
+      masks.push_back("m0");
+      argTypes.push_back(BOOLEAN());
+      argNames.push_back(masks.back());
+    }
+
+    // 10% of times use global aggregation (no grouping keys).
+    std::vector<std::string> groupingKeys;
+    if (vectorFuzzer_.coinToss(0.1)) {
+      ++statistics_.numGlobal;
+    } else {
+      ++statistics_.numGroupBy;
+
+      auto numGroupingKeys =
+          boost::random::uniform_int_distribution<uint32_t>(1, 5)(rng_);
+      for (auto i = 0; i < numGroupingKeys; ++i) {
+        groupingKeys.push_back(fmt::format("g{}", i));
+
+        // Pick random scalar type.
+        argTypes.push_back(vectorFuzzer_.randType(0 /*maxDepth*/));
+        argNames.push_back(groupingKeys.back());
+      }
+    }
+
+    // Generate random input data.
+    auto inputType = ROW(std::move(argNames), std::move(argTypes));
+    std::vector<RowVectorPtr> input;
+    for (auto i = 0; i < 10; ++i) {
+      input.push_back(vectorFuzzer_.fuzzInputRow(inputType));
```

Comment:
We should consider enhancing the vector fuzzer to generate only constant vector for certain parameter, which is a common pattern in aggregate functions.  Otherwise we are wasting quite a lot of the search space just run invalid inputs and compare 2 exceptions.

Reply:
This is a fair point. We do have this as a TODO in the Expression Fuzzer and once we have that we can apply the same to Aggregation Fuzzer. BTW, would you like to help build this capability?